### PR TITLE
Support quoting list values in mappings

### PR DESCRIPTION
**Pull request by Lonami**
## What do these changes do?

These changes resolve aio-libs/aiohttp#4714.

## Are there changes in behavior for the user?

Yes, users can now use lists or tuples when quoting mappings.

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

### CODE REVIEW

**Comment by Lonami** [comment]
I haven't checked performance but if it's an issue it can be improved by re-implementing `_query_var` inside the new `_query_seq_pairs`. And if performance is an absolute concern, we may use the trick where we avoid global look-ups to things like `isinstance` or the `tuple/list` types.

**Comment by Lonami** [comment]
With regards to:
> Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

Is that necessary? Taking a quick look at the commit log, it seems to be used inconsistently anyway (and many commits don't follow that convention).

**Comment by Lonami** [comment]
Oh, I hadn't seen the `CHANGES/.TEMPLATE.rst` file because it starts with a dot so it was hidden. Was I supposed to use that in the changes? If yes, it could be a good idea to rename it to something else and add comments indicating "WRITE HERE" or something.

**Comment by webknjaz** [comment]
> Oh, I hadn't seen the `CHANGES/.TEMPLATE.rst` file because it starts with a dot so it was hidden. Was I supposed to use that in the changes?

No, that template is used by towncrier to generate `CHANGES.rst` (which we run before making releases). You should usually add a short change note file and that's it. Here's an example https://github.com/aio-libs/yarl/blob/01e529690fbb6d122e4a4a6575e1aa5edba4ce47/CHANGES/389.feature, check how it looks in https://github.com/aio-libs/yarl/blob/master/CHANGES.rst#features.

**Comment by webknjaz** [comment]
> Is that necessary? Taking a quick look at the commit log, it seems to be used inconsistently anyway (and many commits don't follow that convention).

It's not about the commit messages but about the changes to docs.

File: tests/test_update_query.py [hunk 1]
```diff
@@ -117,6 +117,45 @@ def test_with_query_list_int():
...
+def test_with_query_sequence_single_then_list():
+    url = URL("http://example.com")
+    expected = URL("http://example.com/?a=1&b=2&b=3")
+    assert url.with_query({"a": 1, "b": [2, 3]}) == expected
+
+
+def test_with_query_sequence_list_then_single():
+    url = URL("http://example.com")
+    expected = URL("http://example.com/?a=1&a=2&b=3")
+    assert url.with_query({"a": [1, 2], "b": 3}) == expected
```
Comment (webknjaz) [inline tests/test_update_query.py:144]: these tests are essentially one sigle test and so it should use `@pytest.mark.parametrize`. You can use `pytest.param(id=)` if you want to improve the test report.

File: tests/test_update_query.py [hunk 2]
```diff
@@ -117,6 +117,45 @@ def test_with_query_list_int():
...
+def test_with_query_sequence_nested_sequence():
+    url = URL("http://example.com")
+    with pytest.raises(TypeError):
+        url.with_query({"a": [[1]]})
+
+
+def test_with_query_sequence_using_pairs():
+    url = URL("http://example.com")
+    with pytest.raises(TypeError):
+        url.with_query([("a", [1, 2])])
```
Comment (webknjaz) [inline tests/test_update_query.py:156]: parametrize plz

File: docs/api.rst [hunk 3]
```diff
@@ -600,6 +608,8 @@ section generates a new *URL* instance.
       URL('http://example.com/path?a=b&c=d')
       >>> URL('http://example.com/path?a=b').update_query({'c': 'd'})
       URL('http://example.com/path?a=b&c=d')
+      >>> URL('http://example.com/path?a=b').update_query({'c': [1, 2]})
+      URL('http://example.com/path?a=b&c=1&c=2')
```
Comment (webknjaz) [inline docs/api.rst:612]: This may be seen as controversial because, for example, web-browsers use `varname[]` format for arrays:
```diff
      URL('http://example.com/path?a=b&c=1&c=2')
      URL('http://example.com/path?a=b&c[]=1&c[]=2')
```

Comment (Lonami) [reply docs/api.rst:612]: Does that not depend on the REST API? Users can use `'c[]'` if the API expects arrays to use `'c[]'` as the key, although I haven't seen that.

Once again, `'c'` and not `'c[]'` is consistent with both `urllib` and `requests`, and sites like Google translate don't use `dt[]` but `dt`.

File: tests/test_update_query.py (hunk 1 above)
Comment (Lonami) [reply tests/test_update_query.py:144]: I'm not that familiar with `pytest`, and none of the previous test used that (although it seems they could), so that's why I didn't. But I will make the changes.

File: tests/test_update_query.py (hunk 2 above)
Comment (Lonami) [reply tests/test_update_query.py:156]: I'm not sure on this one, in my opinion they're quite different in nature so they could deserve different functions, but I will also parametrize it if you want.

File: docs/api.rst (hunk 3 above)
Comment (webknjaz) [reply docs/api.rst:612]: I think this depends on how the backend software handles this. PHP used to handle it gracefully (the form sends `c[]` but then in HTTP handler it's already `c` with values).

You should at least add a good tests coverage including such corner cases.

File: CHANGES/443.feature [hunk 4]
```diff
@@ -0,0 +1,6 @@
+These changes fix https://github.com/aio-libs/aiohttp/issues/4714, that is,
```
Comment (webknjaz) [inline CHANGES/443.feature:1]: This is going to be a part of the changelog under the version.
It'd be weird to see something like:

v1.0.0
======
features
--------
* This changes fix xxx

Comment (webknjaz) [reply CHANGES/443.feature:1]: Because it's not a bug, there shouldn't be a mention of fixing things.
It may say something like "add support for XXX. resolves issue YYY"

File: docs/api.rst (hunk 3 above)
Comment (Lonami) [reply docs/api.rst:612]: Could you suggest a way to test this? We already have tests that cover the fact it's `c` and not `c[]`. I will try to write some test page and server to see what browsers send.

File: tests/test_update_query.py [hunk 5]
```diff
@@ -117,6 +117,34 @@ def test_with_query_list_int():
...
 
 
+@pytest.mark.parametrize(
+    "query,expected",
+    [
+        pytest.param({"a": [1, 2]}, "?a=1&a=2", id="list"),
+        pytest.param({"a": (1, 2)}, "?a=1&a=2", id="tuple"),
+        pytest.param({"a": ["1", 2]}, "?a=1&a=2", id="mixed types"),
+        pytest.param({"a": 1, "b": [2, 3]}, "?a=1&b=2&b=3", id="single then list",),
+        pytest.param({"a": [1, 2], "b": 3}, "?a=1&a=2&b=3", id="list then single",),
```
Comment (webknjaz) [inline tests/test_update_query.py:127]: could also include potentionally buggy cases
```suggestion
        pytest.param({"a": [1, 2], "b": 3}, "?a=1&a=2&b=3", id="list then single"),
        pytest.param({"a": "1&a=2", "b": 3}, "?a=1%26a=2&b=3", id="ampersand value"),
        pytest.param({"a": ["1&a=2", 3], "b": 3}, "?a=1%26a=2&a=3&b=3", id="ampersand value mixed with a list"),
```

File: tests/test_update_query.py [hunk 6]
```diff
@@ -117,6 +117,34 @@ def test_with_query_list_int():
...
+@pytest.mark.parametrize(
+    "query",
+    [
+        pytest.param({"a": [[1]]}, id="nested"),
+        pytest.param([("a", [1, 2])], id="tuple list"),
+    ],
+)
+def test_with_query_sequence_invalid_use(query):
+    url = URL("http://example.com")
+    with pytest.raises(TypeError):
```
Comment (webknjaz) [inline tests/test_update_query.py:144]: the best practice is to always use `match=` arg
```suggestion
    with pytest.raises(TypeError):
```

File: CHANGES/443.feature (hunk 4 above)
Comment (Lonami) [reply CHANGES/443.feature]: I used the fix terminology because it "fixes" an enhancement issue on `aiohttp`, but I've reworded it as "Allow" (I've searched the changelog and there are two other uses of this wording before).

File: docs/api.rst (hunk 3 above)
Comment (webknjaz) [reply docs/api.rst:612]: Just add a test with `{'c[]': [1, 2]}` making sure that `[]` is preserved non-encoded.

File: tests/test_update_query.py (hunk 5 above)
Comment (Lonami) [reply tests/test_update_query.py:127]: I think there's no need to test for `"1&a=2"`, because this test is about the use of sequences, so the latter should be enough. We already cover a similar case in `test_with_query_multidict_with_unsafe` and `test_with_query_params`.

File: CHANGES/443.feature (hunk 4 above)
Comment (webknjaz) [reply CHANGES/443.feature]: Right, things aren't very consistent here, unfortunately. But it shouldn't prevent us from trying to do better :)

File: tests/test_update_query.py (hunk 5 above)
Comment (webknjaz) [reply tests/test_update_query.py:127]: I'd still have it here. Because updates in one part of code can sometimes break others. Better safe than sorry.

File: tests/test_update_query.py (hunk 6 above)
Comment (Lonami) [reply tests/test_update_query.py]: None of the other existing tests do this, so I tried to stay in line with them (and also was unaware of such thing). Made a commit with your suggestion.

File: tests/test_update_query.py (hunk 5 above)
Comment (Lonami) [reply tests/test_update_query.py]: I will update the `test_with_query_params` and `test_with_query_params2` to use `parametrize` too and also add your new case. Does that sound good?

**Comment by Lonami** [comment]
I've also added test for edge cases where the list/tuple are empty or have a single item which we didn't have before. I think it should be ready for review again.

**Comment by Lonami** [comment]
Bumping this now that we're on the weekend. Is there any reason not to merge this other than lack of time?

File: tests/test_update_query.py [hunk 7]
```diff
@@ -117,6 +117,43 @@ def test_with_query_list_int():
     assert str(url.with_query([("a", 1)])) == "http://example.com/?a=1"
 
 
+@pytest.mark.parametrize(
+    "query,expected",
```
Comment (webknjaz) [inline tests/test_update_query.py:121]:
```suggestion
    ("query", "expected"),
```

File: tests/test_update_query.py [hunk 8]
```diff
@@ -117,6 +117,43 @@ def test_with_query_list_int():
...
+        pytest.param({"&": ["=", 2]}, "?%26=%3D&%26=2", id="quote key and value"),
+        pytest.param({"a": 1, "b": [2, 3]}, "?a=1&b=2&b=3", id="single then list"),
+        pytest.param({"a": [1, 2], "b": 3}, "?a=1&a=2&b=3", id="list then single"),
+        pytest.param({"a": ["1&a=2", 3]}, "?a=1%26a%3D2&a=3", id="ampersand then int"),
+        pytest.param({"a": [1, "2&a=3"]}, "?a=1&a=2%26a%3D3", id="int then ampersand"),
+    ],
+)
+def test_with_query_sequence(query, expected):
+    url = URL("http://example.com")
+    assert url.with_query(query) == URL("http://example.com/" + expected)
```
Comment (webknjaz) [inline tests/test_update_query.py:141]: Maybe it'll be better to compare strings? Currently you're testing that the main initializer does the same thing as `URL.with_query()` method.
```suggestion
    assert str(url.with_query(query)) == "http://example.com/{expected}".format_map(locals())
```

File: tests/test_update_query.py [hunk 9]
```diff
@@ -196,16 +233,21 @@ def test_with_query_memoryview():
...
 
-def test_with_query_params():
-    url = URL("http://example.com/get")
-    url2 = url.with_query([("key", "1;2;3")])
-    assert str(url2) == "http://example.com/get?key=1%3B2%3B3"
-
-
-def test_with_query_params2():
+@pytest.mark.parametrize(
+    "query,expected",
```
Comment (webknjaz) [inline tests/test_update_query.py:237]:
```suggestion
    ("query", "expected"),
```

**Comment by webknjaz** [comment]
@Lonami I'd like tests and mentions in all docs for `a[]=1&a[]=2`

**Comment by Lonami** [comment]
Yeah don't mind the extra space there it's fixed in the next commit which I'll push once I polish.

**Comment by Lonami** [comment]
> mentions in all docs for a[]=1&a[]=2

Changing the test to compare `str()` actually did reveal that this escapes `[]` in the key when we shouldn't, although this is a bug from before. That is, this test fails in master:

```python
def test_stuff():
    url = URL("http://example.com/path")
    u2 = url.with_query([("a[]", "3"), ("a[]", "4")])

    assert str(u2) == "http://example.com/path?a[]=3&a[]=4"
```

With:
```
E       AssertionError: assert 'http://examp...D=3&a%5B%5D=4' == 'http://examp...h?a[]=3&a[]=4'
E         - http://example.com/path?a[]=3&a[]=4
E         ?                          ^^    ^^
E         + http://example.com/path?a%5B%5D=3&a%5B%5D=4
E         ?                          ^^^^^^    ^^^^^^

tests/test_update_query.py:228: AssertionError
```

So I suppose it's fine to fix this because people relying on it being escaped was buggy (not intended) behaviour.

**Comment by Lonami** [comment]
Opened #445 to address that.

**Comment by webknjaz** [comment]
As per https://github.com/aio-libs/yarl/pull/445#pullrequestreview-409024181, use encoded values for the square brackets in tests. It is still useful to have tests with non-letter values.

**Comment by Lonami** [comment]
I have "ported" the test using a list of tuples from https://github.com/aio-libs/yarl/pull/445/files#diff-ae226bc0f15cb48ba935d31884c1034dR226 over the new `parametrize` from here, tested also a single key, and fixed the test above to use the quoted braces.

**Comment by Lonami** [comment]
Hooray 🎉, thanks!
